src/services/pallets/PalletsErrorsService.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Substrate API Sidecar.
+//
+// Substrate API Sidecar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import { ApiDecoration } from '@polkadot/api/types';
+import { Text } from '@polkadot/types';
+import {
+	BlockHash,
+	ErrorMetadataLatest,
+	PalletErrorMetadataV14,
+} from '@polkadot/types/interfaces';
+import { IsError } from '@polkadot/types/metadata/decorate/types';
+import { stringCamelCase } from '@polkadot/util';
+import { IPalletErrors, IPalletErrorsItem } from 'src/types/responses';
+
+import { AbstractPalletsService } from '../AbstractPalletsService';
+
+interface IFetchPalletArgs {
+	hash: BlockHash;
+	palletId: string;
+}
+
+interface IFetchErrorItemArgs extends IFetchPalletArgs {
+	errorItemId: string;
+	metadata: boolean;
+}
+
+export class PalletsErrorsService extends AbstractPalletsService {
+	async fetchErrorItem(
+		historicApi: ApiDecoration<'promise'>,
+		{ hash, palletId, errorItemId, metadata }: IFetchErrorItemArgs
+	): Promise<IPalletErrorsItem> {
+		const metadataFieldType = 'errors';
+		const palletMetadata = historicApi.registry.metadata;
+
+		const [palletMeta, palletMetaIdx] = this.findPalletMeta(
```

Comment:
I made a [script](https://github.com/Imod7/testing-sidecar) to test the `errors` endpoint for all the pallets (for example of Kusama when connected to Kusama or Polkadot respectively). The result of the test can be seen in this [screenshot](https://github.com/Imod7/testing-sidecar/blob/main/test-results/errors-kusama.png). It looks like there are 11 pallets for which I get the following error : 
```
{
  "code": 400,
  "message": "\"12\" was not recognized as a queryable pallet.",
  "stack": "BadRequestError: \"12\" was not recognized as a queryable pallet.\n    at PalletsErrorsService.findPalletMeta (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/services/AbstractPalletsService.js:95:19)\n    at PalletsErrorsService.fetchErrors (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/services/pallets/PalletsErrorsService.js:48:50)\n    at PalletsErrorsController.getErrors (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/controllers/pallets/PalletsErrorsController.js:54:75)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/controllers/AbstractController.js:226:9",
  "level": "error"
}
```

This error looks like it is related to the metadata ? 🤔 
I was checking in the current line of code (`line 48`) the `findPalletMeta` call and I was comparing it to the same call in the [storage](https://github.com/paritytech/substrate-api-sidecar/blob/325e80ad0a9e9d8e84e9c138c7dedb92766feaa7/src/services/pallets/PalletsStorageService.ts#L78) endpoint. 
The difference is that 
- in the `storage` endpoint the parameter for the metadata is `chosenMetadata` which is calculated by a call to `chooseMetadataVersion`.
- in this endpoint the `metadata` parameter is a call to `historicApi.registry.metadata` right above (`line 46`). So maybe a call to `historicApi.registry.metadata` is not enough? but a similar calculation as in the `storage` endpoint is needed?

However, I also checked the `storage` endpoint for all the pallets with the same script and the result can be seen in this [screenshot](https://github.com/Imod7/testing-sidecar/blob/main/test-results/storage-kusama.png). It shows 4 pallets for which I get the same error so even though the error appears in less pallets, I still get the error in some of the pallets. So maybe the `chooseMetadataVersion` is a good call to resolve the issue but still needs improvement ? 
cc @TarikGul

Reply:
> in this endpoint the metadata parameter is a call to historicApi.registry.metadata right above (line 46). So maybe a call to historicApi.registry.metadata is not enough? but a similar calculation as in the storage endpoint is needed?

The difference in storage metadata vs pallet/errors is storage has support for V13, whereas for pallets/errors we are only focusing on V14. In fact after @marshacb is completed doing the pallets endpoints we are going to drop support for V13 metadata in storage all together so we wont need `chooseMetadataVersion` anymore.

Reply:
> I made a [script](https://github.com/Imod7/testing-sidecar) to test the `errors` endpoint for all the pallets (for example of Kusama when connected to Kusama or Polkadot respectively). The result of the test can be seen in this [screenshot](https://github.com/Imod7/testing-sidecar/blob/main/test-results/errors-kusama.png). It looks like there are 11 pallets for which I get the following error :
> 
> ```
> {
>   "code": 400,
>   "message": "\"12\" was not recognized as a queryable pallet.",
>   "stack": "BadRequestError: \"12\" was not recognized as a queryable pallet.\n    at PalletsErrorsService.findPalletMeta (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/services/AbstractPalletsService.js:95:19)\n    at PalletsErrorsService.fetchErrors (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/services/pallets/PalletsErrorsService.js:48:50)\n    at PalletsErrorsController.getErrors (/Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/controllers/pallets/PalletsErrorsController.js:54:75)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /Users/tiny_imod/Documents/Parity/substrate-api-sidecar/build/src/controllers/AbstractController.js:226:9",
>   "level": "error"
> }
> ```
> 
> This error looks like it is related to the metadata ? 🤔 I was checking in the current line of code (`line 48`) the `findPalletMeta` call and I was comparing it to the same call in the [storage](https://github.com/paritytech/substrate-api-sidecar/blob/325e80ad0a9e9d8e84e9c138c7dedb92766feaa7/src/services/pallets/PalletsStorageService.ts#L78) endpoint. A difference that drew my attention is that :
> 
> * in the `storage` endpoint the parameter for the metadata is `chosenMetadata` which is calculated by a call to `chooseMetadataVersion`.
> * in this endpoint the `metadata` parameter is a call to `historicApi.registry.metadata` right above (`line 46`). So maybe a call to `historicApi.registry.metadata` is **not** enough? but a similar calculation as in the `storage` endpoint is needed?
> 
> However, I also checked the `storage` endpoint for all the pallets with the same script and the result can be seen in this [screenshot](https://github.com/Imod7/testing-sidecar/blob/main/test-results/storage-kusama.png). It shows 4 pallets for which I get the same error so even though the error appears in less pallets, I still get the error in some of the pallets. So maybe the `chooseMetadataVersion` is a good call to resolve the issue but still needs improvement ? cc @TarikGul ?

@Imod7 I checked the failing `errors` pallets you took a screenshot of and after looking through Kusama/Polkadot codebase I believe those calls fail because no errors are defined or imported to be used by pallets. As far as `chosenMetadata`, as Tarik mentioned we will be removing V13 support in the future so this wont be used or needed going forward.

Reply:
Thank you both so much for the explanations! 🙏 
@marshacb You are right about the failing pallets. It is all good!